packages/abi/src/parser/specifications/v1/resolvable-type.ts

```diff
@@ -0,0 +1,269 @@
+        type.concreteTypeId,
+        undefined
+      );
+    }
+
+    const metadataType = abi.metadataTypes.find(
+      (mt) => mt.metadataTypeId === type.metadataTypeId
+    ) as AbiMetadataTypeV1;
+
+    const concreteTypeArgs = type.typeArguments.map((ta) => {
+      const concreteTypeArg = abi.concreteTypes.find(
+        (ct) => ct.concreteTypeId === ta
+      ) as AbiConcreteTypeV1;
+      return ResolvableType.resolveConcreteType(abi, concreteTypeArg);
+    });
+
+    return new ResolvableType(
+      abi,
+      type.metadataTypeId,
+      ResolvableType.mapTypeParametersAndArgs(abi, metadataType, concreteTypeArgs)
+    ).resolveInternal(type.concreteTypeId, undefined);
+  }
+
+  /**
+   * Metadata types are *handled* and not *resolved* because they might be generic,
+   * in which case they cannot be resolved.
+   * If they're not generic, they can be immediately resolved.
+   */
+  private static handleMetadataType(
+    parent: ResolvableType,
+    abi: AbiSpecificationV1,
+    mt: AbiMetadataTypeV1,
+    typeArguments: AbiComponentV1['typeArguments']
+  ): ResolvableType | ResolvedType {
+    if (swayTypeMatchers.generic(mt.type)) {
+      const resolvedTypeParameter = parent.typeParamsArgsMap?.find(
+        ([tp]) => tp === mt.metadataTypeId
+      )?.[1];
+
+      return resolvedTypeParameter ?? new ResolvableType(abi, mt.metadataTypeId, undefined);
+    }
+
+    if (!mt.components) {
+      /**
+       * types like u8, u16 can make their way into metadata types
+       * if they aren't used _directly_ in a function-input/function-output/log/configurable/messageType
+       * These types are characterized by not having components and we can resolve them as-is
+       */
+      return new ResolvableType(abi, mt.metadataTypeId, undefined).resolveInternal(
+        mt.metadataTypeId,
+        undefined
+      );
+    }
+
+    const typeArgs = typeArguments?.map((ta) => this.handleComponent(parent, abi, ta).type);
+
+    const resolvable = new ResolvableType(
+      abi,
+      mt.metadataTypeId,
+      this.mapTypeParametersAndArgs(abi, mt, typeArgs)
+    );
+
+    if (typeArgs?.every((ta) => ta instanceof ResolvedType)) {
+      return resolvable.resolveInternal(mt.metadataTypeId, undefined);
+    }
+
+    if (resolvable.components?.every((comp) => comp.type instanceof ResolvedType)) {
+      return resolvable.resolveInternal(mt.metadataTypeId, undefined);
+    }
+
+    return resolvable;
+  }
+
+  private resolveInternal(
+    typeId: string | number,
+    typeParamsArgsMap: Array<[number, ResolvedType]> | undefined
+  ): ResolvedType {
+    const typeArgs = this.resolveTypeArgs(typeParamsArgsMap);
+
+    const components: ResolvedType['components'] = this.components?.map((c) => {
+      if (c.type instanceof ResolvedType) {
+        return c as { name: string; type: ResolvedType };
+      }
+
+      const resolvedGenericType = typeArgs?.find(
+        ([tp]) => (c.type as ResolvableType).metadataTypeId === tp
+      )?.[1];
```

Comment:
Why use `.[1];`? We need to properly document this.

Reply:
We're interested in the resolved generic type, which is the second element of the tuple that matches the conditional.

Reply:
I have refactored this whole type params -> args thing into a `Map` which is now a lot simpler.